project: extensibility documentation (#1320)
mia-platform/documentation: docs/console/console-extensibility/registration.md

---

> id: registration

Reviewer: extension-registration ?

title: Register Extensions
sidebar_label: Register Extensions
---
# Register Extensions

> A new extension always should be registered on Console before it can be activated. The registered extension are owned by specific Companies and so the registration and management operations on extensions can only be done by a Company Owner.

Reviewer: ```suggestion
Before [activating an extension](LINK TO ACTIVATION PAGE), you must register it on Console. The registered extension are owned by specific Companies and so the registration and management operations on extensions can only be done by a Company Owner.
```

Author: in my opinion this link on activation only serves to distract the user, because i expect that the user is here to register his extension and could interpret that redirect as something that should be read before to continue with registration

Reviewer: Well this depends on your reading behaviour. If I'm reading this and I am curious about what does it mean to activate an extension I've an immediate link to that.

Also cross-linking is useful for SEO, I'd not miss the opportunity for linking pages together.

## How to register my extension?

If you want to register a new extension, you will need to have already chosen the Company that will own it and ensure that you have the role of Company Owner on that Company. Once this is done, the registration can be done via API, and we recommend using the API Portal to forge requests.

The route to contact is `PUT /api/extensibility/tenants/{tenantId}/extensions`, which you can find under the tags `Companies` or `Extensibility`.

![registration example](img/registrationExample.png)

As can be seen from the example image, registering an extension requires providing some necessary information for proper rendering in the Console.

**Path Params**
- `tenantId`: insert the tenant ID of the Company that will own the new extension

**Body Params**
- `name` (_required_): provide the name of your extension
- `contexts` (_required_): declare in which contexts the extension can be activated. The selectable values are Company and Project.
- `description`: provide a brief description of the extension
- `entry` (_required_): indicate the URL where the iframe will be fetched

> - `extensionType` (_required_): select the type of extension. For now, only the iframe type is supported

Reviewer 2: ```suggestion
- `extensionType` (_required_): select the type of extension. For now, only the `iframe` type is supported
```

as long as `iframe` is the text that we expect to include in the field.

- `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed accoring this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.
- `icon`: select an icon for the new menu item and fill the `icon.name` field. You can find the icons at this [link](https://react-icons.github.io/react-icons/search/), but only Ant, Feather and Phosphor icons are supported. This value can be omitted for the `category` route.
- `labelIntl` (_required_): insert the label to be used on the menu item or menu group and fill this field providing an object JSON `{"en": string, "it": string}` so that the text is also internationalized.
<!-- TODO: Should be added some mentions about the order of menu items? -->

> #### Register Backoffice Extension Example

Reviewer: should this be moved in a tutorial page?

Author: In my opinion this doc pages can be already used as tutorial, because the theorical concepts are not a lot but it is required some practical example near to it to understand them

Reviewer: ok but this is not a tutorial, as it is not clearly stated that the user is following a tutorial. Therefore it's just a doc page with many examples. 

If we want this to be a tutorial let's make one and separate what is documentation from what is a sequence of well defined and exemplified steps the user can follow to fiddle around with the feature.

### How to restrict the extension usage?

A registered extension can specify an array of `permissions` that is used to check whether or not a user can see the extension in the Console, once activated. In particular, the user must have **at least one of the required permissions in the array**. You can find the list of allowed permissions to be registered on an extension in the dedicated table inside the [Identity and access management page](/development_suite/identity-and-access-management/console-levels-and-permission-management.md#identity-capabilities-inside-console).

### How to configure correctly the extension menu item?

![registration example routes](img/registrationExampleRoutes.png)

The routes should include the necessary information to render the extension menu item on the Console sidebar, which enables access to the extension. A route can be registered with the `renderType` left unset to appear as a simple menu item, or it can be a `category` route, creating a new custom menu group to which menu items can be attached.

![menu item and category routes](img/menuItemAndCategory.png)

:::info

> Each extension should have always a single route that is not of type `category` to be meaningful and additionally, it is possible to include also a route of type `category` to add a custom group menu item.

Reviewer 3: ```suggestion
Each extension should have always a route that is not of type `category` to be meaningful and visibile. Additionally, it is possible to include also another route of type `category` to add a custom group menu item and therefore include an extension under that group.
```

Each extension should have at least a route (that is not `category`) to have sense and could also add at most another `category` type route to add a custom group menu item
:::

A route, therefore, requires to provide the following information:

> - `id` (_required_): assign a unique id that can be used for applying overrides with activation or for attaching other menu items if the route renderType is `category`

Reviewer 4: ```suggestion
- `id` (_required_): assign a unique identifier that can be used for applying overrides during activation or for attaching other menu items when the route's `renderType` is set to `category`.
```

- `locationId` (_required_): choose a location to place your route (see the [supported locations](./locations.md))

> - `renderType`: select `category` if you want to add a new menu group, otherwise leave it unsetted to add a menu item

Reviewer 4: ```suggestion
- `renderType`: select `category` if you want to add a new menu group, otherwise leave it unset to add a simple menu item
```

- `parentId`: insert the route `id` of a menu group where you want to attach the menu item. It is possible to choose existing parentIds documented on the [supported locations page](./locations.md) or add a route of type `category` and use its `id`. This value can be omitted for the `category` route.

> - `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed accoring this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.

Reviewer: let's remove the requirement from the api schema

> - `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed accoring this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.

Author: Also for the simple route that are not category? I am afraid that something explode on frontend if the destination path is undefined

> - `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed accoring this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.

Reviewer: Categories should work without `destinationPath`. For now let's keep doc this way but verify nothing breaks and fix the schema and docs asap

> - `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed accoring this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.

Reviewer 4: ```suggestion
- `destinationPath` (_required_): indicate the destination suffix to which you will be redirected when clicking the menu item and it will compose the URL where the iframe will be mounted; in fact, the resulting URL will be composed according to this pattern `<locationPath>/extensions/<extensionId><destinationPath>`. This information should be defined but is ignored for the `category` route.
```

> - `icon`: select an icon for the new menu item and fill the `icon.name` field. You can find the icons at this [link](https://react-icons.github.io/react-icons/search/), but only Ant, Feather and Phosphor icons are supported. This value can be omitted for the `category` route.

Reviewer 3: What should I do once I select an icon from the link?

- `icon`: select an icon for the new menu item and fill the `icon.name` field. You can find the icons at this [link](https://react-icons.github.io/react-icons/search/), but only Ant, Feather and Phosphor icons are supported. This value can be omitted for the `category` route.
- `labelIntl` (_required_): insert the label to be used on the menu item or menu group and fill this field providing an object JSON `{"en": string, "it": string}` so that the text is also internationalized.
- `featureToggles`: TBD
- `order`: TBD

> - `matchExactMountPath`: TBD

Reviewer: don't add this, since they are not currently required to be used

> <!-- TODO: Should be added some mentions about the order of menu items? -->

Reviewer: here yes

#### Register Backoffice Extension Example

The following example shows how to register the Backoffice as a Project extension. To reproduce it, you will need to configure the Backoffice and expose it in such a way that it can be used as an iframe.

> Before to proceed make sure you have:

Reviewer 4: ```suggestion
Before proceeding make sure you have:
```

1. A project reachable externally, otherwise create one by referring to this [guide](../project-configuration/create-a-project.mdx).
2. The `Microfrontend Composer Toolkit` application created and exposed to view the Backoffice frontend. Follow this section of the [guide](../../microfrontend-composer/tutorials/basics#setup-the-microservices) to do so.
3. The response from the Backoffice endpoint can be embedded within an iframe. See this [link](../../development_suite/api-console/api-design/endpoints.md#manage-advanced-endpoint-parameters) for more information.

> Once this is done, you can registrate the Backoffice as extension:

Reviewer 4: ```suggestion
Once this is done, you can register the Backoffice as an extension:
```

**Path Params**
```json
{
  "tenantId": "my-tenant-id"
}
```

**Body Params**
```json
{
  "contexts": ["project"],
  "description": "Extension to integrate Backoffice on Console",
  "entry": "https://<my-domain>/mfe-application/home",
  "extensionType": "iframe",
  "name": "Integrated Backoffice",
  "routes": [
    {
      "destinationPath": "/backoffice",
      "icon": {
        "name": "PiProjectorScreenChartLight"
      },
      "id": "backoffice-route",
      "labelIntl": {
        "en": "Integrated Backoffice",
        "it": "Backoffice integrato"
      },
      "locationId": "project",
      "parentId": "my-menu-group"
    },
    {
      "destinationPath": "/something-ignored",
      "id": "my-menu-group",
      "labelIntl": {
        "en": "My Menu Group",
        "it": "Il mio gruppo menu"
      },
      "locationId": "project",
      "renderType": "category"
    }
  ]
}
```

**Response on success**
```json
{
    "extensionId":"my-extension-id"
}
```

> The extension ID will be required to check the extension with the remaining APIs and can be retrieved using a specific API listed below.

Reviewer 4: ```suggestion
The extension ID will be necessary for working with the extension through the remaining APIs and can be retrieved using a specific API listed below.
```

## Get registered Extensions

> The route `GET /api/extensibility/tenants/{tenantId}/extensions` allows you to query all registered extensions under a specific Company and can be easily used on the API Portal under the tags `Companies` or `Extensibility` having the role of Company Owner on the Company that you would require.

Reviewer 4: ```suggestion
The route `GET /api/extensibility/tenants/{tenantId}/extensions` allows you to retrieve all registered extensions under a specific Company. This functionality is readily accessible via the API Portal under the 'Companies' or 'Extensibility' tags, provided you have the Company Owner role for the selected Company.
```
